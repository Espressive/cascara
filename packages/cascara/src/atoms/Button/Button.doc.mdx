---
title: Button
propTable: Button.js
type: 'atoms'
public: false
---

Accessible Button component that enables users to trigger an action or event. It follows the WAI-ARIA Button Pattern.

## Usage

### Default

```jsx
<Button content='Hello World' />
```

### Fluid

```jsx
<div>
  <Button content='Fluid' fluid />
  <button className='ui primary fluid button' type='button'>
    Hello
  </button>
</div>
```

### Link

```jsx
<Button as='a' content='Link' href='https://google.com' target='_blank' />
```

### Negative

```jsx
<Button content='Negative' outcome='negative' />
```

### Positive

```jsx
<Button content='Positive' outcome='positive' />
```
