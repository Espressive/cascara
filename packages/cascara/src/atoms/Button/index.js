export { default, propTypes as buttonPropTypes } from './Button';
