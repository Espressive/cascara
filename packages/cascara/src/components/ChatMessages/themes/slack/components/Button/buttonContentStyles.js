export const buttonContentStyles = {
  root: ({ props: p, variables: v }) => ({
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    fontSize: v.contentFontSize,
    lineHeight: v.contentLineHeight,

    ...(p.size === 'small' && {
      fontSize: v.sizeSmallContentFontSize,
      lineHeight: v.sizeSmallContentLineHeight,
    }),
  }),
};
