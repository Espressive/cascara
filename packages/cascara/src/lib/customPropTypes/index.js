import json from './json';

export default { json };
