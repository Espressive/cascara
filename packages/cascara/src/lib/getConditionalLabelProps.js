/**
 * Conditionally prepare label props in terms of the isLabeled flag:
 *
 * if true, the label is returned as id so it matches the htmlFor in rendered label tag,
 * otherwise, the label is returned as an aria-label so the node is still accessible
 *
 * @param {String} label The labelfor the node
 * @param {Boolean} isLabeled A flag that specifies if a label tag should be displayed
 * @returns {Object} conditionalLabelProps An object to be dot-spread'ed into the node
 */
export const getConditionalLabelProps = (label, isLabeled) =>
  isLabeled ? { id: label } : { 'aria-label': label };
