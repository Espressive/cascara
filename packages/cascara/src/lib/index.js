export { default as getStatusFromDataLength } from './getStatusFromDataLength';
