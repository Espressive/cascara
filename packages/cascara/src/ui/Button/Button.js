import classNames from 'classnames/bind';
import pt from 'prop-types';
import React, { forwardRef } from 'react';
import { Button as ReakitButton } from 'reakit';
import { getSafeLinkRel } from '../../shared/linkUtils';
import styles from './Button.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /** Can render as a different tag or component */
  as: pt.oneOfType([pt.string, pt.elementType]),
  /** Main content of the button */
  content: pt.string,
  /** Makes the button take the width of the parent container */
  fluid: pt.bool,
  /** Sets the color type of the button to follow the theme brand color */
  isBrandColor: pt.bool,
  /** Indicates the outcome of clicking the button ('positive', 'negative') */
  outcome: pt.oneOf(['positive', 'negative']),
  /** A button can have multiple sizes */
  size: pt.oneOf(['small', 'regular', 'large']),
};

// We are going to need a utility to help us dealing with refs. When a composable `as`
// prop is being used with a React component, we need to pass the ref directly to
// Reakit. Otherwise, we need to use React.forwardRef()

const Button = forwardRef(
  (
    {
      as = 'button',
      content = 'Default Content',
      fluid = false,
      isBrandColor = false,
      outcome,
      size = 'regular',
      ...rest
    },
    ref
  ) => {
    const internalClassName = cx(rest.className, {
      _: true,
      basic: !outcome,
      fluid: fluid,
      large: size === 'large',
      negative: outcome === 'negative',
      positive: outcome === 'positive',
      small: size === 'small',
    });

    return (
      <ReakitButton
        {...rest}
        as={as}
        className={internalClassName}
        ref={ref}
        rel={getSafeLinkRel(rest)}
      >
        {content}
      </ReakitButton>
    );
  }
);
Button.propTypes = propTypes;

export default Button;
