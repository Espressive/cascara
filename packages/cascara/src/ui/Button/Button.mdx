# Button: three MDX files

See if this is a good idea.
