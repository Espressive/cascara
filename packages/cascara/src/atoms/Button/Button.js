import classNames from 'classnames';
import pt from '@espressive/prop-types';
import React, { forwardRef } from 'react';
import { Button as ReakitButton } from 'reakit';
import { InlineIcon } from '@iconify/react';
import { getSafeLinkRel } from '../../lib/linkUtils';
// import styles from './Button.module.scss';

const cx = classNames;

const propTypes = {
  /** Can render as a different tag or component */
  as: pt.oneOfType([pt.string, pt.elementType]),
  /** Main content of the button */
  content: pt.string,
  /** Makes the button take the width of the parent container */
  fluid: pt.string,
  // An SVG Object, which will override the content
  icon: pt.shape({}),
  /** Sets the color type of the button to follow the theme brand color */
  isBrandColor: pt.bool,
  /** Indicates the outcome of clicking the button ('positive', 'negative') */
  outcome: pt.oneOf(['positive', 'negative']),
  /** A button can have multiple sizes */
  size: pt.oneOf(['small', 'regular', 'large']),
};

// We are going to need a utility to help us dealing with refs. When a composable `as`
// prop is being used with a React component, we need to pass the ref directly to
// Reakit. Otherwise, we need to use React.forwardRef()

const Button = forwardRef(
  (
    {
      as = 'button',
      content = 'Default Content',
      fluid = false,
      icon,
      isBrandColor = false,
      outcome,
      size = 'regular',
      ...rest
    },
    ref
  ) => {
    // Leaving this here for when we decide to run our own styles
    // const internalClassName = cx(rest.className, {
    //   _: true,
    //   basic: !outcome,
    //   fluid: fluid,
    //   large: size === 'large',
    //   negative: outcome === 'negative',
    //   positive: outcome === 'positive',
    //   small: size === 'small',
    // });

    const legacyClassname = cx(rest.className, {
      basic: !outcome,
      fluid: fluid,
      icon: icon,
      large: size === 'large',
      negative: outcome === 'negative',
      positive: outcome === 'positive',
      small: size === 'small',
      'ui button': true,
    });

    // FDS-137: use action name for button name if no content is specified
    const buttonText = content || rest.name;

    return (
      <ReakitButton
        {...rest}
        as={as}
        className={legacyClassname}
        ref={ref}
        rel={getSafeLinkRel(rest)}
      >
        {icon ? <InlineIcon icon={icon} /> : buttonText}
      </ReakitButton>
    );
  }
);
Button.propTypes = propTypes;

export default Button;
