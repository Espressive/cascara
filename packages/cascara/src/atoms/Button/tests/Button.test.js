import React from 'react';
import { render, screen } from '@testing-library/react';
import { Link, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import Button from '..';

const buttonContent = 'Default Content';
describe('Button', () => {
  describe('default', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(<Button />).container;
    });

    test('renders without any props', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <button> by default', () => {
      const button = screen.getByText(buttonContent);
      // Make sure the actual DOM element type is BUTTON
      expect(button.tagName).toMatch('BUTTON');
      // Check that we also use the correct type for accessibility
      expect(button).toHaveAttribute('type', 'button');
      // Check that we do not use the redundant "role" attribute when the tag is BUTTON
      expect(button).not.toHaveAttribute('role');
    });
  });

  describe('as', () => {
    describe('<div />', () => {
      test('has correct a11y attributes', () => {
        const props = {
          as: 'div',
        };
        render(<Button {...props} />);

        const button = screen.getByText(buttonContent);

        // Should not have a type anymore, but SHOULD have a role
        expect(button).not.toHaveAttribute('type');
        expect(button).toHaveAttribute('role', 'button');
      });
    });

    describe('<a />', () => {
      test('has correct a11y attributes', () => {
        const props = {
          as: 'a',
          href: 'https://google.com',
        };

        render(<Button {...props} />);

        const button = screen.getByText(buttonContent);

        // Should not have either role or type if we render an anchor
        expect(button).not.toHaveAttribute('type');
        expect(button).not.toHaveAttribute('role');
        expect(button).toHaveAttribute('href', props.href);
        expect(button).not.toHaveAttribute('rel');
      });

      test('has secure rel attribute for external links', () => {
        const props = {
          as: 'a',
          href: 'https://google.com',
          target: '_blank',
        };

        render(<Button {...props} />);

        const button = screen.getByText(buttonContent);

        // Make sure we render the target
        expect(button).toHaveAttribute('target', props.target);
        // Make sure we have a secure rel attribute for a link with a target
        expect(button).toHaveAttribute('rel', 'noopener noreferrer');
      });
    });

    describe('<Component />', () => {
      test('renders a React component', () => {
        const props = {
          as: Link,
          to: '/users',
        };

        const history = createMemoryHistory();

        render(
          <Router history={history}>
            <Button {...props} />
          </Router>
        );

        const button = screen.getByText(buttonContent);

        // Make sure this renders a link (from the Link component)
        expect(button.tagName).toMatch('A');
      });

      test.todo('sets a correct React ref when not rendering as a Component');
    });
  });

  describe('isBrandColor', () => {
    test.todo('set up test for brand color from provider');
  });
});
