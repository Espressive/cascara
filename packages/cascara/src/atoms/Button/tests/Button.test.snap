// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default renders without any props 1`] = `
<div>
  <button
    class="basic ui button"
    type="button"
  >
    Default Content
  </button>
</div>
`;
